src/hotspot/share/gc/shared/workgroup.hpp

```diff
@@ -304,12 +304,39 @@ class WorkGangBarrierSync : public StackObj {
 class SubTasksDone: public CHeapObj<mtInternal> {
   volatile bool* _tasks;
   uint _n_tasks;
-  volatile uint _threads_completed;
 
-  // Set all tasks to unclaimed.
-  void clear();
-
-  void all_tasks_completed_impl(uint n_threads, uint skipped[], size_t skipped_size);
+  // make sure verification logic is run only once
+#ifdef ASSERT
+  volatile bool _verification_done = false;
+#endif
+
+  void all_tasks_completed_impl(uint skipped[], size_t skipped_size) {
+#ifdef ASSERT
```

Comment:
Please keep the definition of the method into the .cpp file. It's too long. You can use the DEBUG_ONLY macro here to not need to define it in non-assert code.

Reply:
Revised.